design/libraries/nodes-and-kinds.md

```diff
@@ -0,0 +1,152 @@
+Nodes and Kinds
+===============
+
+Your IPLD library should have two key types:
+
+1. `Node`;
+2. `Kind`.
+
+`Node` should be an interface -- the membership should be open
+(aka, it should be possible for other packages to implement it).
+
+`Kind` should be an enumeration -- a fixed set of named members,
+which should not be extendable.
+
+
+Kind
+----
+
+`Kind` maps very directly onto the definition of
+[Data Model Kinds](../../data-model-layer/data-model.md#kinds).
+
+`Kind` does not include the Schema layer's concept of "struct", etc.
+
+`Kind` must be an enum, **and not a sum type**.  Attempting to implement
```

Comment:
Based on previous conversation I assume you mean with "enum" something that can be extended and "sum type" something fixed. Then it doesn't make sense to me. I can see the `Node` should be open, but `Kind` is a fixed thing.

To me `Node` is an interface, `Kind` is a fixed list of items that describe the Data Model.

Reply:
No, other way around.  Enum is fixed.  Sumtypes aren't.

Enum has a cardinality of the count of its members.

Sumtype has a cardinality of the sum of the cardinality of whatever its members are... so it can easily become "countable infinity" if any one of the members is.

So,

> To me Node is an interface, Kind is a fixed list of items that describe the Data Model.

^ this is correct.

Reply:
(It **seriously** drives me nuts that the Rust syntax conflates these two things into one name.  They have such very, very, categorically different properties... :imp: :angry: :fire: )

Reply:
I really need to brush my programming language/type theory vocab and write it down for 5 year olds. Or perhaps here it doesn't really matter. `Kind` is a fixed set of things (whatever we call it).

Reply:
... Reading again in the light of the morning, I'm thinking "fixed" is probably not quite the word to disambiguate those two either.  Enum and sum types both have fixed membership in terms of types; it's just that the cardinality of member *values* can easily be infinite in a sum type (it happens as soon as a single thing like 'int' or 'string' is part of one of the sum's member types).

Whereas what the Node type should be is an interface, because that can have a non-fixed and not-known-in-advance number of member *types*.

Reply:
Maybe we actually **should** hammer out a quick type systems vocab primer/glossary for ourselves, here in this repo, just for clarity of reference and ease of linking?  Good idea @vmx !

Perhaps something in the 'concepts' directory.  The 'cardinality' doc over there already kind of gets close to flirting with this.  Expanding on it to talk about sum types versus product types etc could be really useful.

Reply:
Having some proper vocab/glossary would be great. I think I slowly grok what you mean with cardinality in regards to enums and sum types. Thinking of "enums" as what the name suggests: *enumerable* items, hence finite. A sum type can have recursion (that's probably not the right word)/wrap "invitite" (as in all ints) values.

Reply:
Yeah, you've... got me on a roll with this kick in the shorts, actually, so thanks for that :D  I'll probably hoist another PR with some content about this in a day or two.

Reply:
> I really need to brush my programming language/type theory vocab and write it down for 5 year olds.

Don't feel bad, I need to engage a compsci translation layer when trying to absorb some of @warpfork's thoughts. This just argues for needing to be more clear when things get overly academic so we can talk to a broad enough audience. It's probably on us when we see things like this to highlight them as needing further carification.